plugins/flow-visibility/clickhouse-monitor/main.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ClickHouse/clickhouse-go"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// The storage percentage at which the monitor starts to delete old records. By default, if the storage usage is larger than 50%, it starts to delete the old records.
+	threshold = 0.5
+	// The percentage of records in Clickhouse will be deleted when the storage is above threshold.
+	deletePercentage = 0.5
+	// The monitor stops for 3 intervals after a deletion to wait for the Clickhouse MergeTree Engine to release memory.
+	skipRoundsNum = 3
+	// Connection to Clickhouse timeout if if fails for 1 minute.
+	connectionTimeout = time.Minute
+	// Retry connection to Clickhouse every 5 seconds if it fails.
+	connectionWait = 5 * time.Second
+	// Time format for timeInserted
+	timeFormat = "2006-01-02 15:04:05"
+)
+
+var (
+	// The name of the table to store the flow records
+	tableName = os.Getenv("TABLE_NAME")
+	// The names of the materialized views
+	mvNames = strings.Split(os.Getenv("MV_NAMES"), " ")
+	// The namespace of the Clickhouse server
+	namespace = os.Getenv("NAMESPACE")
+	// The clickhouse monitor label
+	monitorLabel = os.Getenv("MONITOR_LABEL")
+)
+
+func main() {
+	// The monitor stops working for several rounds after a deletion
+	// as the release of the memory space for Clickhouse MergeTree engine requires time
+	if !skipRound() {
+		connect, err := connectLoop()
+		if err != nil {
+			klog.ErrorS(err, "Error when connecting to Clickhouse")
+			return
+		}
+		deleted := monitorMemory(connect)
+		if deleted {
+			klog.InfoS("Skip rounds after a successful deletion", "skipRoundsNum", skipRoundsNum)
+		} else {
+			klog.InfoS("Next round will not be skipped", "skipRoundsNum", 0)
+		}
+	}
+}
+
+// Checks the k8s log for the number of rounds to skip.
+// Returns true when the monitor needs to skip more rounds and log the rest number of rounds to skip.
+func skipRound() bool {
+	logString, err := getPodLogs()
```

Comment:
@antoninbas , @yanjunz97 I think that if we move away from CronJob we will need to reconsider this logic, and probably that might not be trivial. I think we should do that, but perhaps as a follow-up to this PR.

Reply:
Sure, if we adopt the cronjob solution, we may use a logic to read cronlog instead of the K8S log.

Reply:
I think this is actually an argument in favor of not using a CronJob in the first place. If your logic needs some state (which is the case here), it is better to just have a long-running Pod.

Reply:
Fine by me... I think doing that will require a bit of refactoring...